File: SlackOff/Models/VideoModels.swift
import Foundation

struct VideoCard: Codable, Identifiable, Hashable {
    let slug: String?
    let title: String?
    let url: String?
    let cover: String?
    let rating: String?
    let remark: String?
    let tag: String?
    let section: String?

    var id: String { slug ?? url ?? title ?? UUID().uuidString }
    var coverURL: URL? { cover.flatMap(URL.init(string:)) }
}

struct HomeResponse: Codable {
    let code: Int?
    let message: String?
    let items: [VideoCard]?
}

struct SearchResponse: Codable {
    let code: Int?
    let message: String?
    let items: [VideoCard]?
}

struct Episode: Codable, Identifiable, Hashable {
    let index: Int?
    let name: String?
    let url: String?
    let dataid: String?

    var id: Int { index ?? 0 }
    var number: Int { index ?? 0 }
    var displayName: String { name ?? "\(number)" }
}

struct VideoDetail: Codable {
    let slug: String?
    let title: String?
    let cover: String?
    let score: String?
    let director: String?
    let writer: String?
    let actor: String?
    let typeName: String?
    let area: String?
    let lang: String?
    let release: String?
    let duration: String?
    let alsoKnown: String?
    let description: String?
    let episodeCount: Int?
    let updateInfo: String?
    let episodes: [Episode]?

    var coverURL: URL? { cover.flatMap(URL.init(string:)) }
}

struct DetailResponse: Codable {
    let code: Int?
    let message: String?
    let detail: VideoDetail?
}

struct Stream: Codable, Hashable {
    let mtype: String?
    let bitrate: Int?
    let title: String?
    let description: String?
    let isVip: Bool?
    let locked: Bool?
    let url: String?
    let headers: [String: String]?

    var streamURL: URL? { url.flatMap(URL.init(string:)) }
}

struct PlayInfo: Codable {
    let slug: String?
    let dataid: String?
    let title: String?
    let episode: Int?
    let quality: String?
    let subtitleUrl: String?
    let streams: [Stream]?
}

struct PlayResponse: Codable {
    let code: Int?
    let message: String?
    let play: PlayInfo?
}

File: android/app/src/main/java/com/slackoff/app/model/VideoModels.kt
package com.slackoff.app.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class VideoCard(
    val slug: String? = null,
    val title: String? = null,
    val url: String? = null,
    val cover: String? = null,
    val rating: String? = null,
    val remark: String? = null,
    val tag: String? = null,
    val section: String? = null
) {
    val id: String get() = slug ?: url ?: title ?: kotlin.random.Random.nextInt().toString()
    val coverUrl: String? get() = cover
}

@Serializable
data class HomeResponse(
    val code: Int? = null,
    val message: String? = null,
    val items: List<VideoCard>? = null
)

@Serializable
data class SearchResponse(
    val code: Int? = null,
    val message: String? = null,
    val items: List<VideoCard>? = null
)

@Serializable
data class Episode(
    val index: Int? = null,
    val name: String? = null,
    val url: String? = null,
    val dataid: String? = null
) {
    val id: Int get() = index ?: 0
    val number: Int get() = index ?: 0
    val displayName: String get() = name ?: "$number"
}

@Serializable
data class VideoDetail(
    val slug: String? = null,
    val title: String? = null,
    val cover: String? = null,
    val score: String? = null,
    val director: String? = null,
    val writer: String? = null,
    val actor: String? = null,
    @SerialName("type_name") val typeName: String? = null,
    val area: String? = null,
    val lang: String? = null,
    val release: String? = null,
    val duration: String? = null,
    @SerialName("also_known") val alsoKnown: String? = null,
    val description: String? = null,
    @SerialName("episode_count") val episodeCount: Int? = null,
    @SerialName("update_info") val updateInfo: String? = null,
    val episodes: List<Episode>? = null
) {
    val coverUrl: String? get() = cover
}

@Serializable
data class DetailResponse(
    val code: Int? = null,
    val message: String? = null,
    val detail: VideoDetail? = null
)

@Serializable
data class Stream(
    val mtype: String? = null,
    val bitrate: Int? = null,
    val title: String? = null,
    val description: String? = null,
    @SerialName("is_vip") val isVip: Boolean? = null,
    val locked: Boolean? = null,
    val url: String? = null,
    val headers: Map<String, String>? = null
) {
    val streamUrl: String? get() = url
}

@Serializable
data class PlayInfo(
    val slug: String? = null,
    val dataid: String? = null,
    val title: String? = null,
    val episode: Int? = null,
    val quality: String? = null,
    @SerialName("subtitle_url") val subtitleUrl: String? = null,
    val streams: List<Stream>? = null
)

@Serializable
data class PlayResponse(
    val code: Int? = null,
    val message: String? = null,
    val play: PlayInfo? = null
)